quantus/helpers/utils.py

```diff
@@ -343,21 +322,265 @@ def get_nr_patches(
-        for axis, _ in enumerate(arr.shape)
-    ]
-    if omit_first_axis:
-        unpad_slice[0] = slice(None)
+
+    assert (
+        len(padded_axes) <= arr.ndim
+    ), "Cannot unpad more axes than array has dimensions"
+
+    if isinstance(pad_width, Sequence):
+        assert len(pad_width) == len(
+            padded_axes
+        ), "pad_width and padded_axes have different lengths"
+        for p in pad_width:
+            if isinstance(p, Tuple):
+                assert len(p) == 2, "Elements in pad_width need to have length 2"
+
+    unpad_slice = []
+    for ax in range(arr.ndim):
+        if ax not in padded_axes:
+            unpad_slice.append(slice(None))
+        elif isinstance(pad_width, int):
+            unpad_slice.append(slice(pad_width, arr.shape[ax] - pad_width))
+        elif isinstance(pad_width[padded_axes.index(ax)], int):
+            unpad_slice.append(
+                slice(
+                    pad_width[padded_axes.index(ax)],
+                    arr.shape[ax] - pad_width[padded_axes.index(ax)],
+                )
+            )
+        else:
+            unpad_slice.append(
+                slice(
+                    pad_width[padded_axes.index(ax)][0],
+                    arr.shape[ax] - pad_width[padded_axes.index(ax)][1],
+                )
+            )
     return arr[tuple(unpad_slice)]
+
+
+def expand_attribution_channel(a_batch: np.ndarray, x_batch: np.ndarray):
+    """Expand additional channel dimension(s) for attributions if needed."""
+    if a_batch.shape[0] != x_batch.shape[0]:
+        raise ValueError(
+            f"a_batch and x_batch must have same number of batches ({a_batch.shape[0]} != {x_batch.shape[0]})"
+        )
+    if a_batch.ndim > x_batch.ndim:
+        raise ValueError(
+            f"a must not have greater ndim than x ({a_batch.ndim} > {x_batch.ndim})"
+        )
+
+    if a_batch.ndim == x_batch.ndim:
+        return a_batch
+    else:
+        attr_axes = infer_attribution_axes(a_batch, x_batch)
+
+        # TODO: infer_attribution_axes currently returns dimensions w/o batch dimension
+        attr_axes = [a + 1 for a in attr_axes]
+        expand_axes = [a for a in range(1, x_batch.ndim) if a not in attr_axes]
+
+        return np.expand_dims(a_batch, axis=tuple(expand_axes))
+
+
+def infer_attribution_axes(a_batch: np.ndarray, x_batch: np.ndarray) -> Sequence[int]:
+    """
+    Infers the axes in x_batch that are covered by a_batch.
+    """
+    if a_batch.shape[0] != x_batch.shape[0]:
+        raise ValueError(
+            f"a_batch and x_batch must have same number of batches ({a_batch.shape[0]} != {x_batch.shape[0]})"
+        )
+
+    if a_batch.ndim > x_batch.ndim:
+        raise ValueError(
+            "Attributions need to have <= dimensions than inputs, but {} > {}".format(
+                a_batch.ndim, x_batch.ndim
+            )
+        )
+
+    # TODO: we currently assume here that the batch axis is not carried into the perturbation functions
```

Comment:
OK.

Reply:
noted this with regard to the upcoming batching pr

Reply:
yep, that's fine although we shouldn't really have TODOs dangling in main branch, OK for this one as this is important as a reminder